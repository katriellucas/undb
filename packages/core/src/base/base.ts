import { and } from '@undb/domain'
import type { Option } from 'oxide.ts'
import type { IUpdateBaseSchema } from './base.schema.js'
import type { BaseSpecification } from './interface.js'
import { WithBaseName } from './specifications/base-name.specification.js'
import type { BaseId, BaseName } from './value-objects/index.js'

export class Base {
  id!: BaseId
  name!: BaseName

  static empty() {
    return new this()
  }

  public update(schema: IUpdateBaseSchema): Option<BaseSpecification> {
    const specs: BaseSpecification[] = []

    if (schema.name) {
      specs.push(WithBaseName.fromString(schema.name))
    }

    return and(...specs)
  }
}
