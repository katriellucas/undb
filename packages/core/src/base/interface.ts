import type { CompositeSpecification, ISpecVisitor } from '@undb/domain'
import type { Base } from './base.js'
import type { WithBaseId } from './specifications/base-id.specification.js'
import type { WithBaseName } from './specifications/base-name.specification.js'
import type { WithBaseQ } from './specifications/base-q.specification.js'

export interface IBaseSpecVisitor extends ISpecVisitor {
  withId(v: WithBaseId): void
  withName(v: WithBaseName): void
  withQ(v: WithBaseQ): void
}

export type BaseSpecification = CompositeSpecification<Base, IBaseSpecVisitor>
