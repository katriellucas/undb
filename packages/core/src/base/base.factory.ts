import { and } from '@undb/domain'
import { Base } from './base.js'
import type { ICreateBaseSchema } from './base.schema.js'
import type { BaseSpecification } from './interface.js'
import { WithBaseId } from './specifications/base-id.specification.js'
import { WithBaseName } from './specifications/base-name.specification.js'
import { BaseId } from './value-objects/base-id.vo.js'

export class BaseFactory {
  static create(...specs: BaseSpecification[]): Base {
    return and(...specs)
      .unwrap()
      .mutate(Base.empty())
      .unwrap()
  }

  static new(input: ICreateBaseSchema): Base {
    return this.create(new WithBaseId(BaseId.fromOrCreate(input.id)), WithBaseName.fromString(input.name))
  }
}
