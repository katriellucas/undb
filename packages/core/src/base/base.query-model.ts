import type { Option } from 'oxide.ts'
import type { IQueryBase } from './base.schema'
import type { BaseSpecification } from './interface'

export interface IBaseQueryModel {
  find(spec: Option<BaseSpecification>): Promise<IQueryBase[]>
  findOneById(id: string): Promise<Option<IQueryBase>>
}
