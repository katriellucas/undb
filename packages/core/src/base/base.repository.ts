import type { Option } from 'oxide.ts'
import type { Base } from './base.js'
import type { BaseSpecification } from './interface.js'

export interface BaseRepository {
  find(spec: BaseSpecification): Promise<Base[]>
  findOneById(id: string): Promise<Option<Base>>

  insert(base: Base): Promise<void>
  updateOneById(id: string, spec: BaseSpecification): Promise<void>
  deleteOneById(id: string): Promise<void>
}
