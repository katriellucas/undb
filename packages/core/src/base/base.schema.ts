import { z } from 'zod'
import { tableIdSchema } from '../table/index.js'
import { baseIdSchema, baseNameSchema } from './value-objects/index.js'

export const queryBase = z.object({
  id: baseIdSchema,
  name: baseNameSchema,
})

export type IQueryBase = z.infer<typeof queryBase>

export const createBaseSchema = z.object({
  id: baseIdSchema.optional(),
  name: baseNameSchema,
})

export type ICreateBaseSchema = z.infer<typeof createBaseSchema>

export const updateBaseSchema = z
  .object({
    name: baseNameSchema,
  })
  .partial()

export type IUpdateBaseSchema = z.infer<typeof updateBaseSchema>

export const moveToBaseSchema = z.object({
  tableId: tableIdSchema,
  baseId: baseIdSchema,
})

export type IMoveToBaseSchema = z.infer<typeof moveToBaseSchema>
