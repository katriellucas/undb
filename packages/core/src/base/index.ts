export * from './base.factory.js'
export * from './base.js'
export * from './base.query-model.js'
export * from './base.repository.js'
export * from './base.schema.js'
export * from './interface.js'
export * from './specifications/index.js'
export * from './value-objects/index.js'
