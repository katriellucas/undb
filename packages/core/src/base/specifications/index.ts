export * from './base-id.specification.js'
export * from './base-name.specification.js'
export * from './base-q.specification.js'
