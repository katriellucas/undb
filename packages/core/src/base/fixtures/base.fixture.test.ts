import { createTestBase } from './base.fixture'

describe('base fixture test', () => {
  test('create test base success', () => {
    const base = createTestBase()

    expect(base).toMatchSnapshot()
  })
})
