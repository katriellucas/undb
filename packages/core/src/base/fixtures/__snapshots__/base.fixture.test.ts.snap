// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`base fixture test > create test base success 1`] = `
Base {
  "id": BaseId {
    "props": {
      "value": "baseId",
    },
  },
  "name": BaseName {
    "props": {
      "value": "name",
    },
  },
}
`;
